lib/phoenix_component/declarative.ex

```diff
@@ -407,6 +419,104 @@ defmodule Phoenix.Component.Declarative do
     """)
   end
 
+  defp validate_attr_values!(slot, name, type, values, line, file) do
+    unless is_list(values) and not Enum.empty?(values) do
+      compile_error!(line, file, """
+      :values must be a non-empty list, got: #{inspect(values)}
+      """)
+    end
+
+    for value <- values do
+      case {type, value} do
+        {_type, nil} ->
+          :ok
+
+        {:any, _value} ->
+          :ok
+
+        {:string, value} when not is_binary(value) ->
+          bad_value!(slot, name, type, value, line, file)
+
+        {:atom, value} when not is_atom(value) ->
+          bad_value!(slot, name, type, value, line, file)
+
+        {:boolean, value} when not is_boolean(value) ->
+          bad_value!(slot, name, type, value, line, file)
+
+        {:integer, value} when not is_integer(value) ->
+          bad_value!(slot, name, type, value, line, file)
+
+        {:float, value} when not is_float(value) ->
+          bad_value!(slot, name, type, value, line, file)
+
+        {:list, value} when not is_list(value) ->
+          bad_value!(slot, name, type, value, line, file)
+
+        {{:struct, mod}, value} when not is_struct(value) ->
+          bad_value!(slot, name, mod, value, line, file)
+
+        {_type, _value} ->
+          :ok
+      end
```

Comment:
You can probably encapsulate the case into something called `valid_value?(type, value)` and use it like this:

 ```
unless valid_value?(type, value) do
  bad_value!(slot, name, mod, value, line, file)
end
```

Then likely reuse it for values, examples, and most likely the `:default` validation too!